stdlib/LinearAlgebra/src/adjtrans.jl

```diff
@@ -267,10 +267,23 @@ Broadcast.broadcast_preserving_zero_d(f, tvs::Union{Number,TransposeAbsVec}...)
 
 ## multiplication *
 
+function _dot_nonrecursive(u, v)
+    lu = length(u)
+    if lu != length(v)
+        throw(DimensionMismatch("first array has length $(lu) which does not match the length of the second, $(length(v))."))
+    end
+    if lu == 0
```

Comment:
Re OffsetArrays support, I think it's enough to check `axes` like this?

```suggestion
function _dot_nonrecursive(u::AbstractVecOrMat, v::AbstractVecOrMat)
    if !(axes(u, 1) == axes(v, 2) && axes(u, 2) == axes(v, 1))
        throw(DimensionMismatch("dimensions of the first array $(axes(u)) and the second array $(axes(v)) are incompatible for a dot product."))
    end
    if isempty(u)
```

I think this is a kind of error that would have thrown if we use here `promote_shape(u', v)` (with a better error message).  Since something like `eachindex(a, b)` already throws an error with something like `eachindex(1:3, OffsetArray(1:3, -2:0))` simply based on `axes`, maybe it makes sense to check `axes` here?

Reply:
Alternatively, we could just do 
```julia
@inbounds sum(u[i]*v[i] for i in eachindex(u, v))
```
However, I'm still not actually convinced it's even desirable to throw an error here. I'm also not convinced we *shouldn't* throw an error, but I will point out that that in the current state of affairs `transpose(u::OffsetVector) * v::Vector` currently has the same behaviour as the implementation in this PR (i.e. no error). I feel like it'd be good at least for 1.5 since it's so near, to just follow the lead of `transpose(u) * v` rather than being over eager about throwing errors. 

```julia
julia> using OffsetArrays

julia> let u = OffsetArray([1,2,3], -1:1), v = [1,2,3]
           transpose(u) * v
       end
14
```

Reply:
I'm rather inclined towards throwing an error, as in the matrix-vector case. If the user really wants to contract an `o::OffsetArray` with a `v::Vector`, it is not complicated to do `OffsetArrays.no_offset_view(o)' * v`. Perhaps `no_offset_view` should be exported.